x-pack/plugins/ml/public/application/services/anomaly_explorer_charts_service.ts

```diff
@@ -0,0 +1,1060 @@
+              (chartType !== CHART_TYPE.EVENT_DISTRIBUTION && value !== null)
+            ) {
+              chartData.push({
+                date: +time,
+                value,
+                entity: filterField,
+              });
+            }
+          });
+        } else {
+          chartData = map(metricData, (value, time) => ({
+            date: +time,
+            value,
+          }));
+        }
+      }
+
+      // Iterate through the anomaly records, adding anomalyScore properties
+      // to the chartData entries for anomalous buckets.
+      const chartDataForPointSearch = getChartDataForPointSearch(chartData, records[0], chartType);
+      each(records, (record) => {
+        // Look for a chart point with the same time as the record.
+        // If none found, insert a point for anomalies due to a gap in the data.
+        const recordTime = record[ML_TIME_FIELD_NAME];
+        let chartPoint = findChartPointForTime(chartDataForPointSearch, recordTime);
+        if (chartPoint === undefined) {
+          chartPoint = { date: recordTime, value: null };
+          chartData.push(chartPoint);
+        }
+        if (chartPoint !== undefined) {
+          chartPoint.anomalyScore = record.record_score;
+
+          if (record.actual !== undefined) {
+            chartPoint.actual = record.actual;
+            chartPoint.typical = record.typical;
+          } else {
+            const causes = get(record, 'causes', []);
+            if (causes.length > 0) {
+              chartPoint.byFieldName = record.by_field_name;
+              chartPoint.numberOfCauses = causes.length;
+              if (causes.length === 1) {
+                // If only a single cause, copy actual and typical values to the top level.
+                const cause = record.causes[0];
+                chartPoint.actual = cause.actual;
+                chartPoint.typical = cause.typical;
+              }
+            }
+          }
+
+          if (record.multi_bucket_impact !== undefined) {
+            chartPoint.multiBucketImpact = record.multi_bucket_impact;
+          }
+        }
+      });
+
+      // Add a scheduledEvents property to any points in the chart data set
+      // which correspond to times of scheduled events for the job.
+      if (scheduledEvents !== undefined) {
+        each(scheduledEvents, (events, time) => {
+          const chartPoint = findChartPointForTime(chartDataForPointSearch, Number(time));
+          if (chartPoint !== undefined) {
+            // Note if the scheduled event coincides with an absence of the underlying metric data,
+            // we don't worry about plotting the event.
+            chartPoint.scheduledEvents = events;
+          }
+        });
+      }
+
+      return chartData;
+    }
+
+    function getChartDataForPointSearch(
+      chartData: ChartPoint[],
+      record: AnomalyRecordDoc,
+      chartType: typeof CHART_TYPE[keyof typeof CHART_TYPE]
```

Comment:
Could use `ChartType` suggested for `explorer_constants.ts`.

Reply:
Updated here [`4c0419c` (#94396)](https://github.com/elastic/kibana/pull/94396/commits/4c0419cb0a7f21e41e0308d0a3e6d3566766022d)